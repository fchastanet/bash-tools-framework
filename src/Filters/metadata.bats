#!/usr/bin/env bash

# shellcheck source=src/batsHeaders.sh
source "$(cd "${BATS_TEST_DIRNAME}/.." && pwd)/batsHeaders.sh"

# shellcheck source=src/Filters/metadata.sh
source "${BATS_TEST_DIRNAME}/metadata.sh"

function Filters::metadata::keepOnlyMetadata { #@test
  run Filters::metadata 0 "${BATS_TEST_DIRNAME}/testsData/metadata.sh"
  assert_output "$(cat "${BATS_TEST_DIRNAME}/testsData/metadata.invert0.expected.txt")"
}

function Filters::metadata::removeOnlyMetadata { #@test
  run Filters::metadata 1 "${BATS_TEST_DIRNAME}/testsData/metadata.sh"
  assert_output "$(cat "${BATS_TEST_DIRNAME}/testsData/metadata.invert1.expected.sh")"
}

function Filters::metadata::defaultInvertValue { #@test
  {
    run Filters::metadata
    assert_output "$(cat "${BATS_TEST_DIRNAME}/testsData/metadata.invert0.expected.txt")"
  } <"${BATS_TEST_DIRNAME}/testsData/metadata.sh"
}
