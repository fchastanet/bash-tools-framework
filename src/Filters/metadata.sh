#!/usr/bin/env bash

# filter the metadata only at the beginning of the file
# we do not want to filter eventual documentation in the
# rest of the script
# @param {bool} invert $1 Invert the sense of matching,
# to select non-matching lines:
# - 0 to keep only metadata
# - 1 to remove metadata and keep the rest of the file
# shellcheck disable=SC2120
Filters::metadata() {
  local invert="${1:-0}"
  shift || true
  awk -v invert="${invert}" '
    BEGIN {
      write=!invert
      headerProcessed=0
    }
    {
      line = $0
      if (/^# (BIN_FILE=|ROOT_DIR_RELATIVE_TO_BIN_DIR=|META_[^=]*=|INCLUDE ).*$/) {
        write=(invert==0) || (headerProcessed==1)
      } else if (/^#!/) {
        write=invert
      } else {
        if (invert == 0) {
          # optimization no need to process more lines
          exit 0
        }
        headerProcessed=1
        write=invert
      }
      if (write==1) {
        print line
      } else {
        next
      }
    }
  ' "$@" || test $? = 1
}
