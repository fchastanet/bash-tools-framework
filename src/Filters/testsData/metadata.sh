#!/usr/bin/env bash
# BIN_FILE=<% ${ROOT_DIR} %>/bin/awkLint
# ROOT_DIR_RELATIVE_TO_BIN_DIR=..

.INCLUDE "${ORIGINAL_TEMPLATE_DIR}/_includes/_header.tpl"

test

true

# BIN_FILE=<% ${ROOT_DIR} %>/bin/awkLint2
# ROOT_DIR_RELATIVE_TO_BIN_DIR=..

otherTest
