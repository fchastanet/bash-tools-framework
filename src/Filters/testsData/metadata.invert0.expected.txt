# BIN_FILE=<% ${ROOT_DIR} %>/bin/awkLint
# ROOT_DIR_RELATIVE_TO_BIN_DIR=..
